### PR TITLE
Basic structure for rake tasks

### PR DESCRIPTION
**Pull request by jjgonecrypto**
- moving to Mongoid from MongoMapper
- basic rake task to grab all subs for users and determine frequencies

### CODE REVIEW

File: app/models/scheme.rb [hunk 1]
```diff
@@ -1,9 +1,9 @@
 class Scheme
   include Mongoid::Document
   
-  field :name,type: String
-  field :amount, type: Float
-  field :currency, type: String
+  field :name, :type => String
+  field :amount, :type => Float
+  field :currency, :type => String
```
Comment (jonstorer) [inline app/models/scheme.rb:6]: this is picky as shit, and 110% does not need to happen, but align like:

``` ruby
field :name,     :type => String
field :amount,   :type => Float
field :currency, :type => String
```

File: app/models/subscription.rb [hunk 2]
```diff
@@ -1,7 +1,32 @@
 class Subscription
- include Mongoid::Document
+  include Mongoid::Document
+  include Mongoid::Timestamps
   
   field :service, :type => String
+  field :frequency
```
Comment (jonstorer) [inline app/models/subscription.rb]: it looks like you're putting two values in this field and I'm not sure why you're doing that.

could it make more sense to have

```
FREQUENCY_OPTIONS = [ "Monthly", "Weekly", "Daily" ]

validates :frequency, :presence => true,
                      :inclusion => { :in => FREQUENCY_OPTIONS }

# defines monthly?, weekly?, daily?
# but is probably overkill for 3 methods
FREQUENCY_OPTIONS.each do |option|
  define_method "#{option.downcase}?" do
    frequency.downcase == option.downcase
  end
end  
```

File: app/models/subscription.rb [hunk 3]
```diff
@@ -1,7 +1,32 @@
...
+  def daily?
+    type.upcase == 'D'
+  end
+
+  def weekly?
+    type.upcase == 'W'
+  end
+
+  private
+  def type
```
Comment (jonstorer) [inline app/models/subscription.rb]: I know i gave you this line, but don't use type it's kinda a reserved word

File: app/models/subscription.rb [hunk 4]
```diff
@@ -1,7 +1,32 @@
...
+    @type ||= frequency.scan(/[a-z]$/i).first
+  end
+
+  private
```
Comment (jonstorer) [inline app/models/subscription.rb]: you don't need to define private twice, it works until you define `protected` or `public`

File: app/models/subscription.rb [hunk 5]
```diff
@@ -1,7 +1,32 @@
...
+  def units
```
Comment (jonstorer) [inline app/models/subscription.rb]: you're not using this anywhere, so why are you coding for it?

File: lib/tasks/daily-email.rake [hunk 6]
```diff
@@ -1,10 +1,19 @@
...
+
 namespace :subtrackr do
    desc "Send out daily email"
    task :dailyemail => :environment do
-     @users = User.where(:subscriptions.service => 'Spot')
      
-     @users.each {|u| 
-     puts 'found ' + u.username
-   }
+     users = User.where(:subscriptions.exists => true, :email.exists => true)
```
Comment (jonstorer) [inline lib/tasks/daily-email.rake:9]: I strongly suggest that you store the frequency in the db strictly as Monthly, weekly, daily so you can query on it.

Comment (jonstorer) [inline lib/tasks/daily-email.rake:9]: also, for a task like this, I suggest skipping mongoid and using mongo ruby driver.  I can help you write that. By skipping Mongoid you can use a cursor to be more protected from timeouts.

There's a bit of code we have to drop in to give you easy access, but let's get this written smarty first, then when we need to optimize

File: lib/tasks/daily-email.rake [hunk 7]
```diff
@@ -1,10 +1,19 @@
...
+    
+     users.each do |u|
```
Comment (jonstorer) [inline lib/tasks/daily-email.rake:11]: `User.where(:subscriptions.exists => true, :email.exists => true, :frequncy => 'Monthly').to_a.each do |user|`

File: lib/tasks/daily-email.rake [hunk 8]
```diff
@@ -1,10 +1,19 @@
...
+       
+       u.subscriptions.each do |s|
```
Comment (jonstorer) [inline lib/tasks/daily-email.rake:13]: `user.subscriptions.each do |subscription|`

File: lib/tasks/daily-email.rake [hunk 9]
```diff
@@ -1,10 +1,19 @@
...
+          puts 'monthly' if s.monthly?
```
Comment (jonstorer) [inline lib/tasks/daily-email.rake:14]: `puts "monthly"`

File: lib/tasks/daily-email.rake (hunk 7 above)
Comment (jonstorer) [inline lib/tasks/daily-email.rake:11]: if you save the frequency in the db, then you don't have to ask if it's monthly. youre only pulling the records you need

File: app/models/subscription.rb (hunk 2 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: i had it written out before, but didn't think to split it into two fields. of course it makes more sense.

File: app/models/subscription.rb (hunk 4 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: cheers

File: app/models/subscription.rb (hunk 2 above)
Comment (jonstorer) [inline app/models/subscription.rb]: IMO it looks like you're storing two values in one field.  Generally, I think that's bad practice. What is your reasoning behind this?

File: app/models/subscription.rb (hunk 5 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: just outlining what i was planning to do. must be the OOP in me.

File: lib/tasks/daily-email.rake (hunk 6 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:9]: interesting....

File: lib/tasks/daily-email.rake (hunk 7 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:11]: agreed, thou

File: app/models/subscription.rb (hunk 5 above)
Comment (jonstorer) [inline app/models/subscription.rb]: the TDDer in me is saying no.  Also, I strongly suggest unit tests. it helps with scope creep and fragile code

File: lib/tasks/daily-email.rake (hunk 7 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:11]: agreed, though i don't want to query against monthly directly. i'll need to write something that says, if time between now and the next due date minus days before is -ve, then add the subscription to a list, and ill email the list to the user.

File: lib/tasks/daily-email.rake (hunk 8 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:13]: haha - yes of course

File: lib/tasks/daily-email.rake (hunk 9 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:14]: single vs double quotes?

File: lib/tasks/daily-email.rake (hunk 7 above)
Comment (jonstorer) [inline lib/tasks/daily-email.rake:11]: mongo does allow you to query against dates, so that might be how to want to do that something like

```
:payment_date => {"$lte" => some_date_time, "$gte" => some_other_date_time}
```

File: lib/tasks/daily-email.rake (hunk 9 above)
Comment (jonstorer) [inline lib/tasks/daily-email.rake:14]: doesn't really matter, it does in some instances, but most of the time it doesn't

File: app/models/subscription.rb (hunk 2 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: i mean, it makes more sense to have two fields. i will change it.

File: app/models/subscription.rb (hunk 5 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: i know, i know - im getting enough familiarity now to be confident enough to write tests. my reasoning against tdd was that i didn't want to spend half my learning time writing tests, even though it is a decent way to learn a language.

File: lib/tasks/daily-email.rake (hunk 7 above)
Comment (jjgonecrypto) [inline lib/tasks/daily-email.rake:11]: yea, that sounds reasonable. i need to think on it a little tmrw, and i'll come up with something and commit it.

**Comment by jonstorer** [comment]
Justin, may I suggest that you update these one at a time? Smaller more granular changes always seem to work better for me.  Plus, you can squash before you merge this into master.

**Comment by jjgonecrypto** [comment]
are you saying you'd rather have many smaller pull requests?

**Comment by jonstorer** [comment]
Oh, no this was a good size. I'm saying don't feel as though you have to update everything before you push again. I'm on github all day reviewing stuff for work and every so often I swing over here to see if you've updated.  

Short version: smaller pushes and more iterations.

**Comment by jjgonecrypto** [comment]
incidentally, the date operations in mongo aren't great. the best thing it can offer is essentially the javascript eval using $where, and there's a performance hit with that (haven't compared it v ruby). however, as the DB structure is less important than the data, i'm thinking to store more info in the db. 

for example, instead of calculating the next subscription dates from the data when running the rake task, I can store the next_bill_date in the DB whenever the subscription saves (see the lastest commit), and I think that is a more suitable approach when using the document db approach, yah?

**Comment by jonstorer** [comment]
As I read your message, I really understanding the importance of user stories. What you're saying makes sense, but honestly, I'm not even sure of the problem you're solving.

I _think_ this task is going to send a "reminder" of sorts like, "Your GitHub bill hits in 5 days" or something like that. And if that's the case, then, I would definitely save the "next bill date" and have your query be something like.

``` ruby
  Subscription.where(:next_bill_date.gte => Time.at(5.days.from_now).at_beginning_of_day,
                     :next_bill_date.lte => Time.at(5.days.from_now).at_end_of_day)
```

How's that sound?

Also, you're "in mongo v in ruby" statement about speed.  Mongo is faster than Ruby for this.  Ruby is kind of a memory hog.  But this is a rake task and therefor runs in the background somewhere. And at this point you're talking about differences in seconds, maybe a minute. So, nothing to worry about ATM, IMO.

File: app/models/subscription.rb [hunk 10]
```diff
@@ -1,7 +1,23 @@
...
+  include Mongoid::Timestamps
   
-  field :service, :type => String
+  before_create :set_next_bill
+  
+  FREQUENCY_UNITS = [ "day", "week", "month", "year" ]
+  
+  field :service,         :type => String
+  field :amount,          :type => Float
+  field :currency,        :type => String,  :default => "usd"
```
Comment (jonstorer) [inline app/models/subscription.rb]: are you planning on opening this up to foreign currencies right off the bat?

File: app/models/subscription.rb [hunk 11]
```diff
@@ -1,7 +1,23 @@
...
+  field :frequency_unit,  :type => String,  :default => "month"
+  field :frequency_amount,:type => Integer, :default => 1
+  field :next_bill,       :type => Date
   
   embedded_in :user
+  
+  protected 
+  def set_next_bill
+    next_bill = Time.new + frequency_amount.send(frequency_unit)
```
Comment (jonstorer) [inline app/models/subscription.rb]: I don't understand what you're trying to do here? Can you expand on the issues and you're method for solving it?

_If_ I understand it correctly, I have an idea, but I'm not sure I do

**Comment by jjgonecrypto** [comment]
yea sorry - i assumed it was self explanatory. you are correct. the rake task is for this story:

"As a user, I want to receive an email reminding me of all services set to rebill me within a specified time frame".

File: app/models/subscription.rb (hunk 11 above)
Comment (jjgonecrypto) [inline app/models/subscription.rb]: i'm trying to set the time for the next bill as a distance from the current time

eg.

```
 next_bill = Time.new + 1.month 
```

or 

```
 next_bill = Time.new + 2.week
```

seems to be working fine for the moment

Comment (jonstorer) [inline app/models/subscription.rb]: I might need you to walk me through it, but I can't get my head around how that would work.  Especially when bills are due at times like "on the 10th day of the month"

and for that, I would try to remove the `Time.now` from the next_bill date.  And the rake task would be something like

``` ruby
#11th day of the month
Subscription.where(:frequency_unit => "monthly", 
                   :next_bill_date.gte => (10.days.to_i),
                   :next_bill_date.lt => (11.days.to_i))
```

And

``` ruby
# 2nd day of a week
Subscription.where(:frequency_unit => "weekly", 
                   :next_bill_date.gte => (1.days.to_i),
                   :next_bill_date.lt => (2.days.to_i))
```

Thoughts? or is this over kill?

By just storing the "Time into a frequency" you may be able to get away from deal with relative time.

Comment (jjgonecrypto) [inline app/models/subscription.rb]: they aren't always. a lot of subscriptions start on the day of the month that you signed up. for those, this is where this will come in handy. for others, i'll will override this functionality with a DOM field, rather than the default `Time.new`

Comment (jonstorer) [inline app/models/subscription.rb]: "The day of the month that you signed up" can be represented as "the Nth of the month".  I would do my best to keep everything the same way. And it looks, to me, that you can do that with the frequency and the day into the frequency. If I'm wrong (or you just wanna do it your way), I'll shut up :-D.  This is your project. :metal:

Comment (jjgonecrypto) [inline app/models/subscription.rb]: but it gets more complicated as you look at different frequencies. 

Monthly starting on day X is easy. So is quarterly, halfly(?) and yearly. The only trick is that if you subscribed on a day that doesn't exist in all months (eg. the 30th or 29th). I imagine some subscriptions deal with those differently (ie. some would charge on Feb 28 for a 30th of the month subscription, and others would do Mar 1). However, if **subtrackr** simply assumes the earlier date, that would make the most sense.   

Daily, weekly are harder as the date changes as the months change. Then you just have to add the increments on from whatever the start date is, regardless of DOM. Before you mention it, no i don't have any daily or weekly subscriptions, but I still want to handle them (some sickness i have i guess).

Comment (jjgonecrypto) [inline app/models/subscription.rb]: anyhow, the problem is really just a data structure, and at the end of the day, I can revist this later. I'd rather focus more on rails, ruby, mongoid issues now while i have your attn =)

Comment (jjgonecrypto) [inline app/models/subscription.rb]: btw, this

```
 Subscription.where(:frequency_unit => "weekly", 
               :next_bill_date.gte => (1.days.to_i),
               :next_bill_date.lt => (2.days.to_i))
```

isn't valid cause it's embedded. it spits out:

```
Mongoid::Errors::InvalidCollection: Access to the collection for Subscription is not allowed since it is an embedded document, please access a collection from the root document.
```

is there a way to do it via `User` that delves into the embedded doc? In Mongoid, it suggests: `person.addresses.where(city: "Berlin")` but that means i need an instance of `user` yah?

Comment (jonstorer) [inline app/models/subscription.rb]: You're right, about the 29th, 30th, and 31st. But those can be handled pretty easily. Like you said, 29, 30, and 31 could be all be handled like 28, or... whatever.

Daily is easy, as it's ... wait for it ... every day.  

Weekly can be easy too.  Because it would all be relative to the beginning of a week and that's it. The actual date wouldn't really matter.

Comment (jonstorer) [inline app/models/subscription.rb]: It looks as though you started using `Subscription` as a Top Level Document and it shouldn't be embedded any more. Which makes sense.  You're really less interested in the user and more interested in specific subscriptions and then their associated user.

Comment (jjgonecrypto) [inline app/models/subscription.rb]: well, i am still interested in the user cause that's whose subscriptions im collating and will need to notify. you think it's better to do subscriptions first and then collate them after based on user?

Comment (jonstorer) [inline app/models/subscription.rb]: exactly, go at it through the subscriptions.

``` ruby
Subscription.where(options).each do |subscription|
  user = subscription.user
  # do some shit with the user
end
```

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -46,6 +46,8 @@ def edit
   def create
     @user = User.new(params[:user])
 
+    @user.subscriptions = [Subscription.new]
```
Comment (jonstorer) [inline app/controllers/users_controller.rb:49]: you shouldn't be creating the subscription here

File: config/routes.rb [hunk 13]
```diff
@@ -1,12 +1,13 @@
 Subtrackr::Application.routes.draw do
   resources :services
 
-  match 'users' => 'users#index', :via => :get
```
Comment (jonstorer) [inline config/routes.rb:4]: whoa

``` ruby
  resources :users do
    resources :subscriptions
  end
```

File: app/controllers/users_controller.rb (hunk 12 above)
Comment (jjgonecrypto) [inline app/controllers/users_controller.rb:49]: sorry, still cleaning up...

File: config/routes.rb (hunk 13 above)
Comment (jjgonecrypto) [inline config/routes.rb:4]: will that work with my users using the `username` in the lookup instead of `id`?

Comment (jonstorer) [inline config/routes.rb:4]: If you add that bit of code I sent you.  You'll need to save the "friendly id" on the record so you can query for it. but it's method chaining. 

 The psuedo code would be...

before Model.find
do        Model.find_with_friendly_id
if no result, Model.find

File: app/controllers/users_controller.rb (hunk 12 above)
Comment (jjgonecrypto) [inline app/controllers/users_controller.rb:49]: i had that in earlier for an example

**Comment by jonstorer** [comment]
Once this bit of code is in, would you be interested in doing a bit of story review and pointing?  This scope of this pull request is HUGE.

What I mean is, the stories I'm used to are:
- A member can sign up - 4pts
- A member can create a subscription - 1 pts
- A member can create a yearly, semi-yearly, quarterly, monthly, semi-monthly, bi-weekly, weekly, and daily subscription - 2pts
- A member can set their 'bill day' - 2pts
- A member frequency and bill date are calculated for simple querying, 4 pts,
- Emails for yearly billings are sent, 4 pts
- Emails for the other frequencies are sent, 4 pts
- Emails contain all reminders that should be sent on that day (ie. yearly + monthly reminders in one email/day, vs multiple/day)

**Comment by jjgonecrypto** [comment]
This is true. I still have unit tests to write and everything. For the next pull request, i'll make it more manageable. I'll come up with some stories now that I'm getting familiar with the environment
